structured/privilege.go

```diff
@@ -0,0 +1,217 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Marc Berhault (marc@cockroachlabs.com)
+
+package structured
+
+import (
+	"fmt"
+	"sort"
+
+	"github.com/cockroachdb/cockroach/security"
+	"github.com/cockroachdb/cockroach/sql/privilege"
+)
+
+// rawToPrivilegeList converts a set of privileges
+// into a PrivilegeList.
+func rawToPrivilegeList(m uint32) privilege.List {
+	ret := privilege.List{}
+	for _, p := range privilege.ByValue {
+		if m&(1<<p) != 0 {
+			ret = append(ret, p)
+		}
+	}
+
+	return ret
+}
+
+// Sort methods for the PrivilegeDescriptor.Users list.
+type userPrivilegeList []*UserPrivileges
+
+func (upl userPrivilegeList) Len() int {
+	return len(upl)
+}
+
+func (upl userPrivilegeList) Swap(i, j int) {
+	upl[i], upl[j] = upl[j], upl[i]
+}
+
+func (upl userPrivilegeList) Less(i, j int) bool {
+	return upl[i].User < upl[j].User
+}
+
+// findUserIndex looks for a given user and returns its
+// index in the User array if found. Returns -1 otherwise.
+func (p *PrivilegeDescriptor) findUserIndex(user string) int {
+	idx := sort.Search(len(p.Users), func(i int) bool {
+		return p.Users[i].User >= user
+	})
+	if idx < len(p.Users) && p.Users[idx].User == user {
+		return idx
+	}
+	return -1
+}
+
+// findUser looks for a specific user in the list.
+// Returns (nil, false) if not found, or (obj, true) if found.
+func (p *PrivilegeDescriptor) findUser(user string) (*UserPrivileges, bool) {
+	idx := p.findUserIndex(user)
+	if idx == -1 {
+		return nil, false
+	}
+	return p.Users[idx], true
+}
+
+// findOrCreateUser looks for a specific user in the list, creating it if needed.
+func (p *PrivilegeDescriptor) findOrCreateUser(user string) *UserPrivileges {
+	idx := sort.Search(len(p.Users), func(i int) bool {
+		return p.Users[i].User >= user
+	})
+	if idx == len(p.Users) {
+		// Not found but should be inserted at the end.
+		p.Users = append(p.Users, &UserPrivileges{User: user})
+	} else if p.Users[idx].User == user {
+		// Found.
+	} else {
+		// New element to be inserted at i.
+		p.Users = append(p.Users, nil)
+		copy(p.Users[idx+1:], p.Users[idx:])
+		p.Users[idx] = &UserPrivileges{User: user}
+	}
+	return p.Users[idx]
+}
+
+// removeUser looks for a given user in the list and removes it if present.
+func (p *PrivilegeDescriptor) removeUser(user string) {
+	idx := p.findUserIndex(user)
+	if idx == -1 {
+		// Not found.
+		return
+	}
+
+	copy(p.Users[idx:], p.Users[idx+1:])
+	p.Users[len(p.Users)-1] = nil
```

Comment:
do you care about order? Because otherwise you can just move the element at the end in place of the one you're deleting: `a[del], a[last] = a[last], nil`